File: AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

/*
Runtime: 2 ms, faster than 99.46% of Java online submissions for Add Two Numbers.
Memory Usage: 42.5 MB, less than 89.13% of Java online submissions for Add Two Numbers.
*/

// Optimized Solution TC: O(n) SC: O(1)
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        
        // Taking Pointers To Start Of Both The Lists
        ListNode l1Pointer = l1;
        ListNode l2Pointer = l2;
        
        // Creating A New Node For Storing The Sum
        ListNode head = new ListNode(0);
        
        // Creating Pointer To The head
        ListNode currentNode = head;
        int carry = 0;
        
        // Iterating Till All 3 Conditions Satisfy
        while(l1Pointer != null || l2Pointer != null || carry !=0){
            
            // Taking Values As Input Till List Hits Null And If The Loop Still Continues To Run THen We Take 0 as Input
            int l1Value = (l1Pointer != null) ? l1Pointer.val : 0;
            int l2Value = (l2Pointer != null) ? l2Pointer.val : 0;
            
            // Calculating Sum
            int sum = l1Value + l2Value + carry;
            
            // Calculating Last Digit
            int lastDigit = sum%10;
            
            // Calculating Carry
            carry = sum/10;
            
            // Storing Last Digit In A Node And Adding It To The Answer
            ListNode lastDigitNode = new ListNode(lastDigit);
            currentNode.next = lastDigitNode;
            currentNode = currentNode.next;
            
            // Base Iterations
            if(l1Pointer != null){
                l1Pointer = l1Pointer.next;
            }
            if(l2Pointer != null){
                l2Pointer = l2Pointer.next;
            }
            
        }
        
        // Returning The Final Result
        return head.next;
        
    }
}

File: AddTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    const result = new ListNode(0);  // Create a dummy head node for the result
    let resultPointer = result;      // Pointer to traverse and build the result list
    let carry = 0;                   // Initialize carry for digit-wise addition

    while (l1 || l2 || carry) {      // Loop while there are digits or a carry-over
        const l1Val = l1 ? l1.val : 0;   // Get the value from l1 (0 if l1 is null)
        const l2Val = l2 ? l2.val : 0;   // Get the value from l2 (0 if l2 is null)
        const sum = l1Val + l2Val + carry; // Calculate the sum with carry

        carry = Math.floor(sum / 10);       // Update the carry (quotient)
        resultPointer.next = new ListNode(sum % 10); // Create a new node with the digit (remainder) 
        resultPointer = resultPointer.next;          // Move the pointer to the new node

        if (l1) l1 = l1.next; // Advance l1 if possible
        if (l2) l2 = l2.next; // Advance l2 if possible
    }

    return result.next; // Return the actual head of the result list (skip the dummy node)
};
